### PR TITLE
feat: implement support for service bindings

### PR DESCRIPTION
**Pull request by threepointone**
This adds experimental support for service bindings, aka worker-to-worker bindings. It's lets you "call" a worker from another worker, without incurring any network cost, and (ideally) with much less latency. To use it, define a `[services]` field in `wrangler.toml`, which is a map of bindings to worker names (and environment). Let's say you already have a worker named "my-worker" deployed. In another worker's configuration, you can create a service binding to it like so:

```toml
[[services]]
binding = "MYWORKER"
service = "my-worker"
environment = "production" # optional, defaults to the worker's `default_environment` for now  
```

And in your worker, you can call it like so:

```js
export default {
  fetch(req, env, ctx) {
    return env.MYWORKER.fetch(new Request("./some-path"));
  },
};
```

### Current critical drawbacks: 

- The api is a little buggy at the moment. We're investigating that internally.
- It connects to, by default, the 'production' worker. If that has a KV or Durable Object or whatever, you could have data loss in production
- Following which, say you run wrangler dev on two bound workers, then the requests do not go the development versions, and it's not obvious
- No local mode support 
- Hard to test

Fixes https://github.com/cloudflare/wrangler2/issues/1026

### CODE REVIEW

File: packages/wrangler/src/dev/local.tsx [hunk 1]
```diff
@@ -87,6 +87,11 @@ function useLocalWorker({
           '⎔ A "public" folder is not yet supported in local mode.'
         );
       }
+      if (Number(bindings.services?.length) > 0) {
```
Comment (JacobMGEvans) [inline packages/wrangler/src/dev/local.tsx:90]: `Number(undefined)` => NaN 
`NaN > 0` is false 
Seems quirky, but works.

File: packages/wrangler/src/worker.ts [hunk 2]
```diff
@@ -157,6 +163,7 @@ export interface CfWorkerInit {
     data_blobs: CfDataBlobBindings | undefined;
     durable_objects: { bindings: CfDurableObject[] } | undefined;
     r2_buckets: CfR2Bucket[] | undefined;
+    services: CfService[] | undefined;
```
Comment (JacobMGEvans) [inline packages/wrangler/src/worker.ts:166]: If there is a default empty array, then undefined would no be necessary 
```suggestion
    services: CfService[] | [];
```
Then the services length check would also not need a coercion. Alternatively it could be checked for `undefined` before that length check and set to empty array.

Comment (threepointone) [reply packages/wrangler/src/worker.ts:166]: I'm not sold on whether the default should an array (including for some other fields) so I'm staying defensive here.

Comment (threepointone) [reply packages/wrangler/src/worker.ts:166]: That our validation automatically sets it so makes me uncomfortable, and I want to revisit it sometime.

Comment (JacobMGEvans) [reply packages/wrangler/src/worker.ts:166]: > I want to revisit it sometime.

Fair enough then 😄

File: .changeset/hot-insects-cry.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+feat: implement support for service bindings
+
+This adds experimental support for service bindings, aka worker-to-worker bindings. It's lets you "call" a worker from another worker, without incurring any network cost, and (ideally) with much less latency. To use it, define a `[services]` field in `wrangler.toml`, which is a map of bindings to worker names (and environment). Let's say you already have a worker named "my-worker" deployed. In another worker's configuration, you can create a service binding to it like so:
+
+```toml
+[[services]]
+binding = "MYWORKER"
+service = "my-worker"
+environment = "production" # optional, defaults to the worker's `default_environment` for now
```
Comment (petebacondarwin) [inline .changeset/hot-insects-cry.md:13]: What exactly is the worker's "default environment" right now??

File: .changeset/hot-insects-cry.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+```
+
+And in your worker, you can call it like so:
+
+```js
+export default {
+  fetch(req, env, ctx) {
+    return env.MYWORKER.fetch(new Request("./some-path"));
```
Comment (petebacondarwin) [inline .changeset/hot-insects-cry.md:21]: Can we actually call a service worker binding with an arbitrary request?
I seem to remember, when playing with this that you had to use the original request, or at least some clone of it.

File: packages/wrangler/src/__tests__/configuration.test.ts [hunk 5]
```diff
@@ -644,6 +645,13 @@ describe("normalizeAndValidateConfig()", () => {
             preview_bucket_name: "R2_PREVIEW_2",
           },
         ],
+        services: [
+          {
+            binding: "SERVICE_BINDING_1",
+            service: "SERVICE_BINDING_1",
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/configuration.test.ts:651]: NIT: I would give the service a different name to its binding, just to be sure we haven't accidentally switched them.

```suggestion
            service: "SERVICE_TYPE_1",
```

File: packages/wrangler/src/__tests__/publish.test.ts [hunk 6]
```diff
@@ -3165,6 +3165,12 @@ addEventListener('fetch', event => {});`
       mockUploadWorkerRequest({
         expectedType: "sw",
         expectedBindings: [
+          { json: 123, name: "ENV_VAR_ONE", type: "json" },
+          {
+            name: "ENV_VAR_TWO",
+            text: "Hello, I'm an environment variable",
+            type: "plain_text",
+          },
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/publish.test.ts:3173]: Out of curiosity (OOC) why did these properties need to move?

File: packages/wrangler/src/config/environment.ts [hunk 7]
```diff
@@ -281,6 +281,18 @@ interface EnvironmentNonInheritable {
     preview_bucket_name?: string;
   }[];
 
+  /**
+   * Specifies service bindings (worker-to-worker) that are bound to this Worker environment.
+   */
```
Comment (petebacondarwin) [inline packages/wrangler/src/config/environment.ts:286]: Should add:

```ts
   * NOTE: This field is not automatically inherited from the top level environment,
   * and so must be specified in every named environment.
   *
   * @default `[]`
   * @nonInheritable
```

File: packages/wrangler/src/config/validation.ts [hunk 8]
```diff
@@ -1192,8 +1187,13 @@ const validateUnsafeBinding: ValidatorFn = (diagnostics, field, value) => {
     const safeBindings = [
       "plain_text",
       "json",
+      "wasm_module",
+      "data_blob",
+      "text_blob",
       "kv_namespace",
       "durable_object_namespace",
+      "r2_bucket",
```
Comment (petebacondarwin) [inline packages/wrangler/src/config/validation.ts:1195]: These new "safe" bindings seem unrelated to this PR?

File: packages/wrangler/src/create-worker-upload-form.ts [hunk 9]
```diff
@@ -32,13 +32,15 @@ export interface WorkerMetadata {
...
+  // If you add any new binding types here, also add it to safeBindings
+  // under validateUnsafeBinding in config/validation.ts
   bindings: (
-    | { type: "kv_namespace"; name: string; namespace_id: string }
     | { type: "plain_text"; name: string; text: string }
     | { type: "json"; name: string; json: unknown }
     | { type: "wasm_module"; name: string; part: string }
     | { type: "text_blob"; name: string; part: string }
     | { type: "data_blob"; name: string; part: string }
+    | { type: "kv_namespace"; name: string; namespace_id: string }
```
Comment (petebacondarwin) [inline packages/wrangler/src/create-worker-upload-form.ts:43]: OOC why move this?

File: packages/wrangler/src/create-worker-upload-form.ts [hunk 10]
```diff
@@ -68,6 +71,14 @@ export function createWorkerUploadForm(worker: CfWorkerInit): FormData {
...
   const metadataBindings: WorkerMetadata["bindings"] = [];
 
+  Object.entries(bindings.vars || {})?.forEach(([key, value]) => {
+    if (typeof value === "string") {
+      metadataBindings.push({ name: key, type: "plain_text", text: value });
+    } else {
+      metadataBindings.push({ name: key, type: "json", json: value });
+    }
+  });
+
```
Comment (petebacondarwin) [inline packages/wrangler/src/create-worker-upload-form.ts:81]: OOC why move this?

File: packages/wrangler/src/index.tsx [hunk 11]
```diff
@@ -892,6 +892,14 @@ export async function main(argv: string[]): Promise<void> {
       const config = readConfig(configPath, args);
       const entry = await getEntry(args, config, "dev");
 
+      if (config.services && config.services.length > 0) {
+        logger.warn(
+          `This worker is bound to live services: ${config.services
+            .map((service) => `${service.binding} (${service.service})`)
```
Comment (petebacondarwin) [inline packages/wrangler/src/index.tsx:898]: I think we should also show the environment here too.

File: packages/wrangler/src/__tests__/dev.test.tsx [hunk 12]
```diff
@@ -839,6 +839,34 @@ describe("wrangler dev", () => {
...
+
+
+        [33m▲ [43;33m[[43;30mWARNING[43;33m][0m [1mThis worker is bound to live services: WorkerA (A), WorkerB (B)[0m
+
+        ",
+        }
+      `);
+    });
+  });
 });
```
Comment (petebacondarwin) [inline packages/wrangler/src/__tests__/dev.test.tsx:870]: Can we add a test that shows the error when using service bindings in `--local` mode?

**Review by petebacondarwin — APPROVED** [review]
A bunch of NITs and suggestions. Nothing blocking landing this. Feel free to fix up and you see fit.

File: packages/wrangler/src/config/validation.ts (hunk 8 above)
Comment (threepointone) [reply packages/wrangler/src/config/validation.ts:1195]: added it because it was missed when it was actually added, it simply adds a warning if you use them in unsafe.bindings.

File: packages/wrangler/src/create-worker-upload-form.ts (hunk 9 above)
Comment (threepointone) [reply packages/wrangler/src/create-worker-upload-form.ts:43]: just aesthetics, felt it belonged with the other data primitive types

File: packages/wrangler/src/create-worker-upload-form.ts (hunk 10 above)
Comment (threepointone) [reply packages/wrangler/src/create-worker-upload-form.ts:81]: just aesthetics, like above. vars, then data storage, etc. nothing functional.

File: packages/wrangler/src/__tests__/dev.test.tsx (hunk 12 above)
Comment (threepointone) [reply packages/wrangler/src/__tests__/dev.test.tsx:870]: I'll follow up on this PR for this

**Comment by Plausible0Bug** [comment]
I can confirm the above syntax works however the one described in your documentation is not?
https://github.com/cloudflare/cloudflare-docs/blob/f3937ef122fd9b82220aeca8014e0c114fb4c3eb/content/workers/wrangler/configuration.md?plain=1#L217

I'm publishing my project using [cloudflare/wrangler-action@2.0.0](https://github.com/cloudflare/wrangler-action).
